File: src/app.controller.ts
import { Body, Controller, Param, DefaultValuePipe, ParseIntPipe, Post, Inject, Logger, Query } from "@nestjs/common";
import { ApiTags } from "@nestjs/swagger";
import { ExampleProviderService } from "./app.service";
import {
  FeedValuesRequest,
  FeedValuesResponse,
  FeedVolumesResponse,
  RoundFeedValuesResponse,
} from "./dto/provider-requests.dto";

@ApiTags("Feed Value Provider API")
@Controller()
export class ExampleProviderController {
  private logger = new Logger(ExampleProviderController.name);

  constructor(@Inject("EXAMPLE_PROVIDER_SERVICE") private readonly providerService: ExampleProviderService) {}

  /**
   * Retrieves feed values for a specific voting round.
   * Used by FTSOv2 Scaling clients.
   * @param votingRoundId - The ID of the voting round.
   * @param body - The request body containing the feed IDs.
   * @returns The feed values for the specified voting round.
   */
  @Post("feed-values/:votingRoundId")
  async getFeedValues(
    @Param("votingRoundId", ParseIntPipe) votingRoundId: number,
    @Body() body: FeedValuesRequest
  ): Promise<RoundFeedValuesResponse> {
    const values = await this.providerService.getValues(body.feeds);
    this.logger.log(`Feed values for voting round ${votingRoundId}: ${JSON.stringify(values)}`);
    return {
      votingRoundId,
      data: values,
    };
  }

  /**
   * Retrieves the latest feed values without a specific voting round ID.
   * Used by FTSOv2 Fast Updates clients.
   * @param body - The request body containing the feed IDs.
   * @returns The latest feed values.
   */
  @Post("feed-values/")
  async getCurrentFeedValues(@Body() body: FeedValuesRequest): Promise<FeedValuesResponse> {
    const values = await this.providerService.getValues(body.feeds);
    this.logger.log(`Current feed values: ${JSON.stringify(values)}`);
    return {
      data: values,
    };
  }

  /**
   * Retrieves the trading volumes for the requested feeds over a specified time window.
   * @param body - The request body containing the feed IDs.
   * @param windowSec - The time window in seconds for which to retrieve the volume. Defaults to 60.
   * @returns The trading volumes for the requested feeds.
   */
  @Post("volumes/")
  async getFeedVolumes(
    @Body() body: FeedValuesRequest,
    @Query("window", new DefaultValuePipe("60"), ParseIntPipe) windowSec: number
  ): Promise<FeedVolumesResponse> {
    const values = await this.providerService.getVolumes(body.feeds, windowSec);
    this.logger.log(`Feed volumes for last ${windowSec} seconds: ${JSON.stringify(values)}`);
    return {
      data: values,
    };
  }
}

File: src/app_controller.py
from fastapi import Body, Query, Path
from pynest.common import Controller, Post, Inject, DefaultValuePipe, ParseIntPipe
from loguru import logger
from typing import Annotated

from src.app_service import AppService
from src.dto.provider_requests import (
    FeedValuesRequest,
    FeedValuesResponse,
    FeedVolumesResponse,
    RoundFeedValuesResponse,
)


@Controller()
class AppController:
    @Inject()
    def __init__(self, app_service: AppService):
        self.app_service = app_service
        self.logger = logger

    @Post("feed-values/{voting_round_id}")
    async def get_feed_values(
        self,
        voting_round_id: Annotated[int, Path(alias="voting_round_id")],
        body: FeedValuesRequest = Body(...),
    ) -> RoundFeedValuesResponse:
        values = await self.app_service.get_values(body.feeds)
        self.logger.info(f"Feed values for voting round {voting_round_id}: {values}")
        return RoundFeedValuesResponse(votingRoundId=voting_round_id, data=values)

    @Post("feed-values")
    async def get_current_feed_values(self, body: FeedValuesRequest = Body(...)) -> FeedValuesResponse:
        values = await self.app_service.get_values(body.feeds)
        self.logger.info(f"Current feed values: {values}")
        return FeedValuesResponse(data=values)

    @Post("volumes")
    async def get_feed_volumes(
        self,
        body: FeedValuesRequest = Body(...),
        window_sec: int = Query(60, alias="window"),
    ) -> FeedVolumesResponse:
        values = await self.app_service.get_volumes(body.feeds, window_sec)
        self.logger.info(f"Feed volumes for last {window_sec} seconds: {values}")
        return FeedVolumesResponse(data=values)